src/Features/Core/Portable/EmbeddedLanguages/AbstractLanguageDetector.cs

```diff
@@ -171,16 +173,48 @@ private bool IsEmbeddedLanguageStringLiteralToken(SyntaxToken token, SemanticMod
         }
     }
 
-    internal abstract class AbstractLanguageDetector<TOptions, TTree> :
+    internal interface ILanguageDetectorInfo<TDetector>
+    {
+        ImmutableArray<string> LanguageIdentifiers { get; }
+        TDetector Create(Compilation compilation, EmbeddedLanguageInfo info);
+    }
```

Comment:
this pattern allows the derived types to just specify thsi information, and teh base type can take care of everything else.